qcodes/instrument/channel.py

```diff
@@ -529,3 +532,183 @@ def validate(self, value, context: str='') -> None:
+                channel is accessible
+            name: channel name
+            existence (bool): True if the channel exists on the instrument
+            channel_list (N52xxChannelList): If the channel is deleted,
+                delete `self` from the channel list
+        """
+        super().__init__(parent, name=name, **kwargs)
+        self._exists_on_instrument = existence
+        self._channel_list = channel_list
+
+    def create(self) ->None:
+        """Create the channel on the instrument"""
+        self._create()
+        self._exists_on_instrument = True
+
+    def _create(self) ->None:
+        """
+        (SCPI) commands needed to create the instrument. Note that we
+        need to use `self.root_instrument.write` to send commands,
+        because self.write will cause _assert_existence to raise a
+        runtime error.
+        """
+        raise NotImplementedError("Please subclass")
+
+    def delete(self) ->None:
+        """
+        Delete the channel from the instrument and remove from channel list
+        """
+        self._delete()
+        if self._channel_list is not None:
+            self._channel_list.remove(self)
+        self._exists_on_instrument = False
+
+    def _delete(self) ->None:
+        raise NotImplementedError("Please subclass")
+
+    def _assert_existence(self) ->None:
+        if not self._exists_on_instrument:
+            raise RuntimeError(
+                "Object does not exist (anymore) on the instrument")
+
+    def write(self, cmd: str) ->Any:
+        """
+        Only write to the instrument if the channel is present on the instrument
+        """
+        self._assert_existence()
+        return super().write(cmd)
+
+    def ask(self, cmd: str) ->None:
+        """
+        Only query to the instrument if the channel is present on the instrument
+        """
+        self._assert_existence()
+        return super().ask(cmd)
+
+    @property
+    def exists_on_instrument(self):
+        return self._exists_on_instrument
+
+
+class AutoLoadableChannelList(ChannelList):
+    """
+    Extends the QCoDeS ChannelList class to add the following features:
+    - Automatically create channel objects on initialization
+    - Make a `add` method to create channel objects
+    """
+    def __init__(
+            self,
+            parent: Instrument,
+            name: str,
+            chan_type: type,
+            chan_list: Optional[Sequence['AutoLoadableInstrumentChannel']]=None,
+            snapshotable: bool=True,
+            multichan_paramclass: type=MultiChannelInstrumentParameter,
+            **kwargs
+    ) ->None:
+
+        super().__init__(
+            parent, name, chan_type, chan_list, snapshotable,
+            multichan_paramclass
+        )
+
+        new_channels = self._chan_type.load_from_instrument(
+            self._parent, channel_list=self, **kwargs)
+
+        for channel in new_channels:
+            # NB: There is a bug in `extend`. TODO: Make a PR!
```

Comment:
what is the bug about? Are you planning to make a PR with a test and a fix? That'd be great to do before landing this PR.

Reply:
Ok, I will make a PR for that as well. Yes, there is a bug I discovered in the `ChannelList` class, but I have not come round to fixing this and making a PR. There is something wrong with the `extend` method in the `ChannelList` class, but I forgot what, exactly :-)